src/Microsoft.VisualStudio.Composition/Configuration/Formatters/MetadataDictionaryFormatter.cs

```diff
@@ -0,0 +1,363 @@
+                        messagePackWriter.Write((byte)ObjectType.CreationPolicy);
+                        messagePackWriter.Write((byte)(CreationPolicy)value);
+                        break;
+
+                    case Type objectType when typeof(Type).GetTypeInfo().IsAssignableFrom(objectType):
+                        TypeRef typeRefValue = TypeRef.Get((Type)value, options.CompositionResolver());
+                        messagePackWriter.Write((byte)ObjectType.Type);
+                        options.Resolver.GetFormatterWithVerify<TypeRef?>().Serialize(ref messagePackWriter, typeRefValue, options);
+                        break;
+
+                    case Type objectType when objectType == typeof(TypeRef):
+                        messagePackWriter.Write((byte)ObjectType.TypeRef);
+                        options.Resolver.GetFormatterWithVerify<TypeRef>().Serialize(ref messagePackWriter, (TypeRef)value, options);
+                        break;
+
+                    case Type objectType when typeof(LazyMetadataWrapper.Enum32Substitution) == objectType:
+                        var enum32SubstitutionValue = (LazyMetadataWrapper.Enum32Substitution)value;
+                        messagePackWriter.Write((byte)ObjectType.Enum32Substitution);
+                        options.Resolver.GetFormatterWithVerify<TypeRef?>().Serialize(ref messagePackWriter, enum32SubstitutionValue.EnumType, options);
+                        options.Resolver.GetFormatterWithVerify<int?>().Serialize(ref messagePackWriter, enum32SubstitutionValue.RawValue, options);
+                        break;
+
+                    case Type objectType when typeof(LazyMetadataWrapper.TypeSubstitution) == objectType:
+                        var typeSubstitutionValue = (LazyMetadataWrapper.TypeSubstitution)value;
+                        messagePackWriter.Write((byte)ObjectType.TypeSubstitution);
+                        options.Resolver.GetFormatterWithVerify<TypeRef?>().Serialize(ref messagePackWriter, typeSubstitutionValue.TypeRef, options);
+                        break;
+
+                    case Type objectType when typeof(LazyMetadataWrapper.TypeArraySubstitution) == objectType:
+                        var typeArraySubstitutionValue = (LazyMetadataWrapper.TypeArraySubstitution)value;
+                        messagePackWriter.Write((byte)ObjectType.TypeArraySubstitution);
+                        options.Resolver.GetFormatterWithVerify<IReadOnlyCollection<TypeRef?>>().Serialize(ref messagePackWriter, typeArraySubstitutionValue.TypeRefArray, options);
+
+                        break;
+
+                    default:
+                        var typeLessData = MessagePackSerializer.Typeless.Serialize(value);
+                        messagePackWriter.Write((byte)ObjectType.TypeLess);
+                        options.Resolver.GetFormatterWithVerify<byte[]>().Serialize(ref messagePackWriter, typeLessData, options);
```

Comment:
A formatter should _not_ call back into the toplevel `MessagePackSerializer.Serialize` method. Instead, it should call directly into a formatter. So something like this:

```cs
messagePackWriter.WriteArrayHeader(2);
messagePackWriter.Write((byte)ObjectType.Typeless);
TypelessFormatter.Instance.Serialize(ref writer, value, options);
```